tests/lambdas/leverDev.test.js

```diff
@@ -0,0 +1,79 @@
+const lambda = require('../../src/functions/lever')
+const got = require('got')
+const url = require('url')
+
+jest.mock('got')
+jest.mock('url')
+
+describe('different data, not production', () => {
```

Comment:
Following on from [this](https://github.com/yldio/yld.io/pull/329/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.yml#r313721201) comment. It is worth taking a look at the documentation from jest to understand how describe blocks are intended to be used, please read these two links: 

- [Describe](https://jestjs.io/docs/en/api#describename-fn)
- [Test](https://jestjs.io/docs/en/api#testname-fn-timeout) - this is an alias of `it` which we are using in our tests. 

What I've created in my branch is something that looks like this:
```
describe(lever lambda)
    afterEach(() => {
     jest.clearAllMocks()
   })
    it('should call the lever webhook and return X when there are differences', () => { // test // })
    it('should only return X and not call lever webhook when there are no differences', () =>  { // test // })
```

This set up means that we have a much better way of composing each unit test for our function that we want to test. Mocks are set in the individual unit test (`it(...)`) and then reset after each test has finished running.

Instead of running multiple describe blocks for the same function we're isolating it to one block and making it a lot more concise. Another benefit of this is understanding the output from lever when we are writing tests and also looking at failed test reports in CI

```
PASS  tests/lambdas/lever.test.js
  Lever
    ✓ fires a request to the LAMBDA_LEVER_WEBHOOK url when there are differences in jobs (6ms)
    ✓ returns 200 and no difference in body (1ms)
````

Whereas the set up in this branch would be

```
FAIL tests/lambdas/leverDev.test.js
   different data, not production
     ✕ returns 200 and no difference in body
```

Reply:
This tests pass both locally and on CI. The only reason that that I have two different files is because I couldn't figure out how to pass different env.vars for faking prod vs faking dev. 

My bad though because I just copied-pasted the `it` description for the leverDev, and the test is actually testing what it's meant to test.
![image](https://user-images.githubusercontent.com/31079643/63012248-d4195780-be81-11e9-9829-d13dddf76511.png)

Reply:
The way you've done it has 4 `expect` tests in one `it` - I think we're thinking in different levels of testing - to me, it feels each `expect`  be inside a separate `it` so that each gets a separate tick as the tests run. 

Would you always test in pairs, so to speak: that `got` was called with the values we expect, and that the `response` is what we expect it to be? If not, what is the deciding factor as to what gets `expected`?

Reply:
Each `it` block is a scenario you want to test, in my branch we're testing two scenarios:

- `fires a request to the LAMBDA_LEVER_WEBHOOK url when there are differences in jobs`
- `returns 200 and no difference in body`

We set up the mocks return values and other various requirements then we execute the function. We want to make sure that our function operates correctly and to make sure a function fires correctly we have to check several things. It's not against the rules to use multiple `expect`s in each testing block. One `it` is a **scenario**, for that scenario to pass we have to assert something or somethings. For a simple function that can be one thing but in this case we need to make sure that we're doing several things correctly.

Reply:
What do you mean testing in pairs? A very common pattern in unit tests is `response` (from the function under test) and `expected` (what we expect the function to return). 

The deciding factor of what to test should be fairly self evident if you read through the function you want to test. In the lever function we want to make sure that:
- The URLs are parsed correctly, to do this we `expect got to be called with X` for each url
- If there are difference we want got.post to be called and with the right argument, to do this we `expect got.post to be called with X`
- Then we want to make sure the responses are correct so we just set up `expected` and compare this with the return value of the function.

It's important to understand the core aim of the function, here it's "make sure that the lever webhook url is requested if there are differences in job IDs" and "this is how this happens". Understanding this is then how we decide what to assert on with `expect`

Reply:
re prod and dev, i'm not too concerned about the dev instance as it only occurs locally, thus if there are issues when developing them you'll be aware and then need to fix them.